### PR TITLE
return errors for some extra DHT issues

### PR DESCRIPTION
**Pull request by ljedrz**
Return explicit subscription errors on some of the more user-facing DHT requests, namely `Bootstrap`, `StartProviding` and `PutRecord`; `Kademlia` errors for those are currently only logged and `KadResult::Complete` is returned, potentially concealing issues.

### CODE REVIEW

File: src/p2p/behaviour.rs [hunk 1]
```diff
@@ -104,7 +106,14 @@ impl<Types: IpfsTypes> NetworkBehaviourEventProcess<KademliaEvent> for Behaviour
                         );
                     }
                     Bootstrap(Err(BootstrapError::Timeout { .. })) => {
-                        warn!("kad: failed to bootstrap");
+                        warn!("kad: timed out while trying to bootstrap");
+
+                        if self.kademlia.query(&id).is_none() {
```
Comment (koivunej) [inline src/p2p/behaviour.rs:111]: Just to clarify, this:

```
if self.kademlia.query(&id).is_none() {
```

Is done to to guard against the query returning a result but still being ongoing or something?

**Review by koivunej — APPROVED** [review]
I am not sure how good idea it's to always log an error before propagating it but perhaps here is no other choice, since we don't really know if the other side will log it, or if there is even something ongoing.

But yeah, seems that things are only moving towards better errors and more straightforward processing so approve.

File: src/p2p/behaviour.rs (hunk 1 above)
Comment (ljedrz) [reply src/p2p/behaviour.rs:111]: yes, exactly that - some queries return partial results, and we don't want to finish the subscription until it's complete

**Pull request merged**